### PR TITLE
[WIP][rllib] Define whether a model outputs logits or probs

### LINKED ISSUE #8532: [rllib] Add an option to define whether a model outputs logits or probs
I propose adding an option to define what `inputs` represent (logits or probs) when using default action distributions. As I understand, there is an assumption that models always return logits, but this is problematic when trying to suppress specific actions by assigning them a zero probability. A common practice (e.g. in [Attention Is All You Need](https://arxiv.org/pdf/1706.03762.pdf), OpenAI Five) is to set logits to -inf, and then compute softmax on that. This however leads to probs being outputted. When we now take these probs and put into e.g. `TorchMultiCategorical` action distribution, the results will be normalized once again and this will break our idea. If we, on the other hand, output logits with -inf values, this will lead to nans when computing entropy.

I know there is an option to define custom action distribution, and I did that, but this leads to changing only a single line in a given action distribution, e.g.:

```python
class TorchMultiCategorical(TorchDistributionWrapper):
    """MultiCategorical distribution for MultiDiscrete action spaces."""

    @override(TorchDistributionWrapper)
    def __init__(self, inputs, model, input_lens):
        super().__init__(inputs, model)
        # If input_lens is np.ndarray or list, force-make it a tuple.
        inputs_split = self.inputs.split(tuple(input_lens), dim=1)
        self.cats = [
            torch.distributions.categorical.Categorical(logits=input_)  # <===== HERE
            for input_ in inputs_split
        ]
```

My idea is to have a property, e.g. in `model`, which is set based on a config or explicitly in a model class, which tells, whether this model outputs logits or probs.

I am fine preparing PR for this (at least for the PyTorch classes), but I'd like to hear your opinion first. :smiley:

### PR DESCRIPTION
**Pull request by iamhatesz**
## Why are these changes needed?

Current action distributions expect models to output logits values, but there are multiple use cases, when directly outputting probs values is more natural. See the original post in #8532.

@sven1977 Could you please take a look and help with TF version? I tried experimenting with `tfp.distributions.Categorical`, but without success (tests for TF fails, when PyTorch runs fine, but I am completely newbie in TF).

For the next steps I'd like to handle this flag also in auto-generated models (`FullyConnectedNetwork`, etc.).

## Related issue number

Closes #8532 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.
